pymess/backend/dialer/__init__.py

```diff
@@ -0,0 +1,116 @@
+        relation
+        :param tag: string mark that will be saved with the message
+        :param template: template object from which content of the message was created
+        :param kwargs: extra attributes that will be saved with the message
+        """
+        try:
+            message = DialerMessage.objects.create(
+                recipient=recipient,
+                content=content,
+                state=self.get_initial_dialer_state(recipient),
+                extra_data=kwargs,
+                tag=tag,
+                template=template,
+                template_slug=template.slug if template else None,
+                **self._get_extra_message_kwargs()
+            )
+            if related_objects:
+                message.related_objects.create_from_related_objects(*related_objects)
+            return message
+        except PersistenceException as ex:
+            raise DialerSendingError(force_text(ex))
+
+    def publish_message(self, message):
+        """
+        Place for implementation logic of sending message.
+        :param message: message instance
+        """
+        raise NotImplementedError
+
+    def get_initial_dialer_state(self, recipient):
+        """
+        returns initial state for logged dialer instance.
+        :param recipient: phone number of the recipient
+        """
+        return DialerMessage.STATE.UNKNOWN
+
+    def _update_dialer_states(self, messages):
+        """
+        If dialer sender provides check dialer delivery this method can be overridden.
+        :param messages: messages which state will be updated
+        """
+        raise NotImplementedError('Check dialer state is not supported with the backend')
+
+    def bulk_check_dialer_states(self):
+        """
+        Method that find messages that are not in the final state and updates their states.
+        """
+        messages_to_check = DialerMessage.objects.exclude(state=DialerMessage.STATE.DONE)
+        if messages_to_check.exists():
+            self._update_dialer_states(messages_to_check)
+
+
+def send_template(recipient, slug, context_data, related_objects=None, tag=None):
+    """
+    Helper for building and sending dialer message from a template.
+    :param recipient: phone number of the recipient
+    :param slug: slug of a dialer template
+    :param context_data: dict of data that will be sent to the template renderer
+    :param related_objects: list of related objects that will be linked with the dialer message using generic
+        relation
+    :param tag: string mark that will be saved with the message
+    :return: dialer message object or None if template cannot be sent
+    """
+    return get_dialer_template_model().objects.get(slug=slug).send(
+        recipient,
+        context_data,
+        related_objects=related_objects,
+        tag=tag
+    )
+
+
+def send(recipient, content, related_objects=None, tag=None, **dialer_kwargs):
```

Comment:
Tato a předchozí metoda jsou duplikace kódu z ostatních bakendů, nejde to generalizovat?

Reply:
pokusim sa

Reply:
snad sa mi to podarilo...